appcelerator/titanium_mobile
# [TIMOB-26015] Android/iOS: Refactored "app.js" launch handling

**JIRA:**
- https://jira.appcelerator.org/browse/TIMOB-26015
- https://jira.appcelerator.org/browse/TIMOB-26144
- https://jira.appcelerator.org/browse/TIMOB-26128
- https://jira.appcelerator.org/browse/TIMOB-26124
- https://jira.appcelerator.org/browse/TIMOB-25899
- https://jira.appcelerator.org/browse/TIMOB-26217

**Summary:**
- Added a new "titanium_mobile/common/Resources" JavaScript folder.
  * Provides JS files/extensions intended to be shared by across all platforms.
- Android/iOS now launch using new `ti.main.js` instead of `app.js`.
  * Used to load JS extension/bootstrap scripts that are shared on all platforms.
  * Loads `app.js` once finished loading everything.
- Now logs app version and Titanium version on startup via `ti.main.js` instead of from native code.
  * New Message: `<AppName> <AppVer> (Powered by Titanium <TiVer>.<TiBuildHash>)`
- [TIMOB-26015](https://jira.appcelerator.org/browse/TIMOB-26015) Added bootstrap script support.
  * All files named `bootstrap.js` or end with name `*.bootstrap.js` under the "Resources" directory will be automatically required-in before `app.js` gets loaded. This is a recursive search.
  * This is the CommonJS equivalent of native Android module's `onAppCreate()` or iOS module's `load()` method.
  * Bootstrap script can support optional `exports.execute(finishedCallback)` method intended to display UI to end-user. Will not load `app.js` until given callback is invoked.
- [TIMOB-26124](https://jira.appcelerator.org/browse/TIMOB-26124) Fixed bug where `Ti.buildHash` always returned `null` on Android as of Titanium 6.0.0.
  * Was a typo in `scons-build.js` and `scons-cleanbuild.js` scripts.
  * Fix needed by new Titanium version logging on app startup. (Build hash needed to differentiate between RC and GA versions.)
- [TIMOB-26128](https://jira.appcelerator.org/browse/TIMOB-26128) Modified Android build system to include asset directory names have a leading underscore '_'.
  * This was an intentional limitation in Google's `aapt` Android command line tool.
  * Solved by using `appt` command line argument `--ignore-assets` and omitting `<dir>_*` from filter list.
- [TIMOB-25899](https://jira.appcelerator.org/browse/TIMOB-25899) Modified JS `Error` object to expose non-enumerable properties via `JSON.stringify()`.
  * Was always returning `"{}"` on Android since none of the V8 `Error` object properties are enumerable.
  * Also modified to reveal `stack` property on iOS. (Not enumerable in JavaScriptCore.)
  * Solved by adding `toJSON()` method to `Error` object prototype. (See: `./common/Resources/ti.internal/extensions/Error.js`)
  * Implementing this was not needed for this PR, but it's a proof of concept on how JS prototype can be extended on all platforms in Titanium's core JS code.
- [TIMOB-26144](https://jira.appcelerator.org/browse/TIMOB-26144) Significantly improved Android's performance in accessing files/directories under the "Resources" directory.
  * The `File.getDirectoryListing()` performance is now about 20x faster. The `isDirectory()`, `isFile()`, and `existence()` calls have been improved too.
  * Now stores a hash table of all file/directory paths under the APK's "assets" folder for quick existence checks. Paths are now obtained by accessing APK as a zip file and fetching its zip entries, which is significantly faster than `AssetManager.list()` calls.
- [TIMOB-26217](https://jira.appcelerator.org/browse/TIMOB-26217) iOS: Added missing `Ti.Filesystem.File` methods `isFile()` and `isDirectory()` when referencing an encrypted asset.
  * Was causing JS runtime errors in bootstrap recursive search.
- [TIMOB-16678](https://jira.appcelerator.org/browse/TIMOB-16678) Modified Android `File.read()` to not return an empty string when referencing an encrypted asset.
  * Needed to access new `bootstrap.json` for device/production builds.

**Notes:**
- I purposely did not document the new `*.bootstrap.js` handling in our API docs. We should dog-food it first to see how much we like it before setting it in stone. Plus we may want it to use `Promises` in the future (this feature is missing on iOS 9 at the moment).
- We plan on having our `ti.playservices` module take advantage of the bootstrapping feature in 7.4.0. It's needed to check/install/update "Google Play Services" before loading the `app.js`.

**Test:**
Perform the below tests on the following:
- Android device without LiveView _(This tests encrypted asset handling.)_
- Android emulator _(This tests APK's "assets" handling.)_
- iOS device
- iOS emulator
- iOS 8 _(This tests oldest JavaScript engine version handling.)_

Test procedure:
1. Build and run the app with the below `app.js` and `ui.bootstrap.js` code. Note that the `ui.bootstrap.js` needs to be next to the `app.js` file under the `Resources` directory.
2. On app startup, verify that an alert is displayed stating that it was launched by a bootstrap.
3. Tap on the "OK" button.
4. Next the `app.js` will have loaded, displaying a JavaScript `Error` object in JSON form. _(This verifies that our core JS extensions have been loaded.)_
5. If running on iOS, verify that the displayed JSON contains a "stack" entry. (This used to be excluded on iOS.)
6. If running on Android, verify that the displayed JSON is not an empty `"{}"`. The JSON should display entries "message" and "stack".
7. If running on Android, press the "Back" button and then restart the app. Verify that the bootstrap alert dialog (mentioned in step 2) is shown again.

---
**app.js**
```javascript
var window = Ti.UI.createWindow({ fullscreen: true });
var scrollView = Ti.UI.createScrollView({
	layout: "vertical",
	showHorizontalScrollIndicator: false,
	shorVerticalScrollIndicator: true,
	width: Ti.UI.FILL,
	height: Ti.UI.FILL,
});
scrollView.add(Ti.UI.createLabel({
	text: "JSON.stringify(new Error('My error message'))",
	width: Ti.UI.FILL,
	top: "10dp",
}));
scrollView.add(Ti.UI.createLabel({
	text: JSON.stringify(new Error('My error message'), null, 4),
	width: "100%",
}));
window.add(scrollView);
window.open();
```

---
**ui.bootstrap.js**
```javascript
exports.execute = function(finished) {
	var dialog = Ti.UI.createAlertDialog({
		title: 'Alert Dialog',
		message: 'This dialog was launched by a bootstrap script.',
		buttonNames: ['OK'],
		cancel: 0,
	});
	dialog.addEventListener('click', function(e) {
		finished();
	});
	dialog.show();
};
```
